pkg/services/ngalert/notifier/multiorg_alertmanager.go

```diff
@@ -104,9 +115,21 @@ func NewMultiOrgAlertmanager(cfg *setting.Cfg, configStore AlertingStore, orgSto
 		ns:            ns,
 		peer:          &NilPeer{},
 	}
+
 	if err := moa.setupClustering(cfg); err != nil {
 		return nil, err
 	}
+
+	// Set up the default per tenant Alertmanager factory.
+	moa.factory = func(ctx context.Context, orgID int64) (Alertmanager, error) {
```

Comment:
Why are we using this pattern instead of just taking an `orgAlertmanagerFactory` in `NewMultiorgAlertmanager()` and using it in case it's not `nil`? Do we plan on having other options applied to the MOA?

Reply:
The main benefit is avoiding a [shotgun surgery](https://refactoring.guru/smells/shotgun-surgery), Initially, I went with what you described, but I had to find and modify every single reference (code + tests) to pass `nil` as is last argument.

IMO, this is a well-known pattern in Go so even though I don't have any other options right now, it might save the next person the hassle. There's also an argument against the list of arguments for this particular method being too long already and introducing a [parameter struct](https://refactoring.guru/introduce-parameter-object) is the way too go - but it ended up having the same problem.

If you feel strongly about it I can file an issue to follow up on it or do it as part of this PR.